### ISSUE TITLE
10-15x performance opportunity for ImmutableRoaringBitmap

### ISSUE THREAD

**Issue by krisskross** [lemire/RoaringBitmap#34]
Hi

I noticed that RoaringBitmap is several thousands times faster than ImmutableRoaringBitmap and got curious as to why this is. 

I wrote a few JMH tests and ran them in mission control and so far I managed to increase the speed by 10-15x by doing the following.

1) Skip the clone in MutableRoaringArray.appendCopy.

2) buffer.asShortBuffer().slice() in ImmutableRoaringArray.getContainerAtIndex steals a lot of performance. Things speed up quite a lot by creating ShortBuffer once and providing it to getContainerAtIndex instead. 

The slice also steal performance but that requires a bigger change I suppose. I cannot prove it but I suspect that a regular ByteBuffer might be faster. Or even better, using Martin Thompson's [DirectBuffer](https://github.com/real-logic/Agrona/blob/master/src/main/java/uk/co/real_logic/agrona/concurrent/UnsafeBuffer.java).

Since i'm unfamiliar with the code i'm not sure if these changes breaks some internals? Seems workable from the tests I made.

Cheers,
-Kristoffer

**Comment by lemire** [comment]
@krisskross  I am sorry , I was away for a few days. Let me look at it.

**Comment by lemire** [comment]
> I noticed that RoaringBitmap is several thousands times faster than ImmutableRoaringBitmap

We do not have tests where that is the case. If you could contribute such a benchmark to the jmh directory of the project, it would be great!

> 1) Skip the clone in MutableRoaringArray.appendCopy.

You refer to this piece of code : 

https://github.com/lemire/RoaringBitmap/blob/master/src/main/java/org/roaringbitmap/buffer/MutableRoaringArray.java#L190-L199

This seems pretty much equivalent to the RoaringBitmap version : 

https://github.com/lemire/RoaringBitmap/blob/master/src/main/java/org/roaringbitmap/RoaringArray.java#L38-L42

Why would one be "thousands of times" slower?

In any case, one can certainly find a way to avoid these cloning calls. The Go version uses a copy-on-write that delays the clone. 

> 2) buffer.asShortBuffer().slice() in ImmutableRoaringArray.getContainerAtIndex steals a lot of performance. Things speed up quite a lot by creating ShortBuffer once and providing it to getContainerAtIndex instead. 

If it does improve performance, then it would be a great contribution.

Maybe it would be a good starting point?

> The slice also steal performance but that requires a bigger change I suppose. 

We could rewrite the code to avoid a slice call... but then you would need to track the location of the content in some other way.

> I cannot prove it but I suspect that a regular ByteBuffer might be faster.

If so, then it would not be very hard to switch to a ByteBuffer.

> Or even better, using Martin Thompson's DirectBuffer.

He uses the Apache license, so there is no problem in using his code...

**Comment by krisskross** [comment]
Sorry, I meant to say +1000 ops/s faster.  Glad to hear you're open for a contribution. I'm trying to enjoy the summer, but I'll see if I can find some time the coming weeks.

**Comment by lemire** [comment]
@krisskross 

What is the point of open source without contributions? ;-)

**Comment by krisskross** [comment]
True :-)

**Comment by lemire** [comment]
@krisskross  Any progress on this front?

**Comment by krisskross** [comment]
Unfortunately not, and not sure when. Maybe close now and reopen later?

**Comment by lemire** [comment]
Ok. I will close and open a new issue about object creation.

**Comment by lemire** [comment]
See https://github.com/lemire/RoaringBitmap/issues/40

I think that it is your main point. 

My main problem is that I do not know how to intelligently benchmark such problems. Yes, there are many tools, but I don't know how to use them well.

Ideally, I'd like a tool that tells me... "this line you have there generated 10,000 of your 100,000 temporary objects during the course of the computation".

That is, I need a way to approach this problem so that I can quantify the problems and then benchmark the changes.

**Comment by bobymicroby** [comment]
@lemire You can use [Java flight recorder](http://hirt.se/blog/?p=381)

**Comment by krisskross** [comment]
I found it very useful to run jmc with jmh and output a file that can inspected after the test is done.

```bash
java -jar target/benchmarks.jar org.TestClass -jvmArgs="-XX:+UnlockCommercialFeatures -XX:+FlightRecorder -XX:StartFlightRecording=duration=60s,filename=file.jfr"
```

**Comment by bobymicroby** [comment]
@krisskross  I do the same thing as well. :+1:

**Comment by lemire** [comment]
@krisskross @bobymicroby 

Sadly, the JVM on my test machine does not appear to support these features.

Could any one of you help by running these two problematic cases? 

          $ cd RoaringBitmap
          $ git pull
          $ cd jmh
          $ ./run.sh needwork.SlowIterate
          $ ./run.sh needwork.SlowORaggregate

These two "needwork" benchmarks are cases where I think that we should be able to double the speed or more. However, we current lack a good diagnosis of what is holding down the performance.

Any help would be appreciated!

**Comment by krisskross** [comment]
I'll have a look later tonight!

**Comment by lemire** [comment]
@krisskross 

Thanks. The SlowIterate test is of particular interest in this context as I strongly suspect it is GC-related... given that I have reviewed the code (there is no much of it) several times for performance issues and see none... other than the fact that we create temporary objects left and right.

But, really, to me, both "needwork" tests are mysterious. I would expect both of them to run much faster. As in 2x, 3x faster...

**Comment by krisskross** [comment]
Here are test runs for SlowORaggregate and SlowIterate. I haven't looked closely at them yet.

https://dl.dropboxusercontent.com/u/39196712/SlowORaggregate.jfr
https://dl.dropboxusercontent.com/u/39196712/SlowIterate.jfr

Btw, what JDK are you using? jmc should be available from 1.7 and onward under $JAVA_HOME/bin/jmc

**Comment by lemire** [comment]
@krisskross 

My problem was with the -XX:+UnlockCommercialFeatures flag which fails on my Linux box.

In any case, thanks for the files. We are investigating possible speed-ups.
